src/workerd/api/streams/internal.c++

```diff
@@ -1162,6 +1162,20 @@ jsg::Promise<void> WritableStreamInternalController::doAbort(
 
   KJ_IF_SOME(writable, state.tryGet<Writable>()) {
     auto exception = js.exceptionToKj(js.v8Ref(reason));
+
+    if (FeatureFlags::get(js).getInternalWritableStreamAbortClearsQueue()) {
+      // If this flag is set, we will clear the queue proactively and immediately
+      // error the stream rather than handling the abort lazily. In this case, the
+      // stream will be put into an errored state immediately after draining the
+      // queue. All pending writes and other operations in the queue will be rejected
+      // immediately and an immediately resolved or rejected promise will be returned.
+      writable->abort(kj::cp(exception), WritableStreamSink::AbortOption::DRAIN);
```

Comment:
If there is currently a `write()` call in progress (that is, the `Promise` returned by a previous `write()` has not resolved yet), then this is UB, at least in the case of `EncodedAsyncOutputStream::abort()` which doesn't make any attempt to cancel that promise before deleting the underlying KJ stream.

One option is to change `WritableStreamSink`'s contract to say it is responsible for canceling writes on abort(). Another option would be to have `WritableStreamInternalController` itself be responsible for wrapping each write promise in a `kj::Canceler` which would be canceled here, before calling `abort()` on the sink. The latter seems better to me since it's handled in one place rather than every sink implementation.

Note that similarly, any `pumpTo()` needs to be canceled before aborting.

Reply:
Yep, was thinking along the same lines... I agree a kj::Canceler here would be a good approach.

Reply:
Ok, updated but needs a careful review to make sure the new behavior is correct. It does not appear that we have any tests that verify the edge cases on this. I'll be looking as seeing if we can reasonably expand the test coverage here. But, in the meantime please take a look